### DISCUSSION TITLE
Optimized the prompt word module of the plugin

### DISCUSSION THREAD

**Discussion by dkqjhoifh** [Acly/krita-ai-diffusion#867] [Ideas]
Krita is the best free ai plugin I've ever used, but its prompts are still only available in English, which is a pain in the way I keep going to the web page to translate prompts during the drawing process, so why not use Argos Translate CLIP Text Encode Node instead of clip prompt? That way, the whole world won't have to worry about prompts and translations.

**Comment by dkqjhoifh** [comment]
![微信截图_20240624173943](https://github.com/Acly/krita-ai-diffusion/assets/171809048/6ac1561d-9b53-4005-9a85-3198c4606fee)

**Comment by dkqjhoifh** [comment]
I tried to manually modify the node of the clip prompt in the krita workflow, although it works in comfyui, but it doesn't seem to take effect in krita, krita still only recognizes English.
![微信截图_20240624174348](https://github.com/Acly/krita-ai-diffusion/assets/171809048/7e9a9e62-5175-44c5-9b83-7ec944335b19)

**Reply by Drakosha405** [reply]
> I tried to manually modify the node of the clip prompt in the krita workflow, although it works in comfyui, but it doesn't seem to take effect in krita, krita still only recognizes English.

I tried to insert this node into the workflow, I didn't find any problems, I changed it to German for verification, and everything is fine in the process logs.
Line 243 in comfy_workflow.py 
`return self.add("ArgosTranslateCLIPTextEncodeNode", 1, clip=clip, text=text, from_translate="ru")`

**Reply by dkqjhoifh** [reply]
So, does your krita already support German prompts?

**Reply by dkqjhoifh** [reply]
Which krita workflow are you modifying and can you tell me the file path?

**Reply by Drakosha405** [reply]
Line 243 in comfy_workflow.py
![image](https://github.com/Acly/krita-ai-diffusion/assets/146111805/7c0e9c33-f2a8-4103-a447-43440de0732f)
dog from German
![image](https://github.com/Acly/krita-ai-diffusion/assets/146111805/308dea20-0ddd-4fdc-a014-08e2824a17e4)
cat from Russian
![image](https://github.com/Acly/krita-ai-diffusion/assets/146111805/97743f23-6143-44a2-884c-1d534d48f746)

**Reply by XylitolJ** [reply]
> > I tried to manually modify the node of the clip prompt in the krita workflow, although it works in comfyui, but it doesn't seem to take effect in krita, krita still only recognizes English.
> 
> I tried to insert this node into the workflow, I didn't find any problems, I changed it to German for verification, and everything is fine in the process logs. Line 243 in comfy_workflow.py `return self.add("ArgosTranslateCLIPTextEncodeNode", 1, clip=clip, text=text, from_translate="ru")`

Unfortunately mine doesn't work, could you please take a look ?
I've already install Alexpet Node:
![image](https://github.com/Acly/krita-ai-diffusion/assets/16879990/7374effc-b330-4f93-b17d-ebff77ccae47)

then insert line like this:

![image](https://github.com/Acly/krita-ai-diffusion/assets/16879990/e6ed449c-5cc8-4b7b-9d34-3ffacb7fe4ac)

but it appear an error when execute
![image](https://github.com/Acly/krita-ai-diffusion/assets/16879990/cc219979-b90d-4e66-ac41-2bd58aa8dcab)

My Krita version: 1.19.0
![image](https://github.com/Acly/krita-ai-diffusion/assets/16879990/a3287868-fa8d-4d4f-8ab6-26f7bfdfc5cc)

in client.log:

`
2024-07-02 15:00:16,701 INFO Found LoRA model hyper for SD 1.5: Hyper-SD15-8steps-CFG-lora.safetensors
2024-07-02 15:00:16,701 INFO Found LoRA model hyper for SD XL: Hyper-SDXL-8steps-CFG-lora.safetensors
2024-07-02 15:00:16,701 INFO Found LoRA model lcm for SD 1.5: lcm-lora-sdv1-5.safetensors
2024-07-02 15:00:16,701 INFO Found LoRA model lcm for SD XL: lcm-lora-sdxl.safetensors
2024-07-02 15:00:16,701 INFO Found LoRA model face for SD 1.5: ipadapter\ip-adapter-faceid-plus_sd15_lora.safetensors
2024-07-02 15:00:16,701 INFO Found LoRA model face for SD XL: ip-adapter-faceid-plusv2_sdxl_lora.safetensors
2024-07-02 15:00:16,701 INFO Found LoRA model lightning for SD XL: sdxl_lightning_8step_lora.safetensors
2024-07-02 15:00:17,733 INFO SD 1.5: supported
2024-07-02 15:00:17,733 INFO SD XL: supported
2024-07-02 15:00:17,733 INFO SD 3: supported
2024-07-02 15:00:33,584 ERROR Job 6b6215e3-2075-4468-80f5-08a6e4ec1d67 failed: 'ru'
['  File "D:\\ComfyUI_windows_portable\\ComfyUI\\execution.py", line 151, in recursive_execute\n    output_data, output_ui = get_output_data(obj, input_data_all)\n                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n', '  File "D:\\ComfyUI_windows_portable\\ComfyUI\\execution.py", line 81, in get_output_data\n    return_values = map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True)\n                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n', '  File "D:\\ComfyUI_windows_portable\\ComfyUI\\execution.py", line 74, in map_node_over_list\n    results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))\n                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n', '  File "D:\\ComfyUI_windows_portable\\ComfyUI\\custom_nodes\\ComfyUI_Custom_Nodes_AlekPet\\ArgosTranslateNode\\argos_translate_node.py", line 182, in argos_translate_text\n    self.langs_support = ALL_CODES[from_translate]["targets"]\n                         ~~~~~~~~~^^^^^^^^^^^^^^^^\n']
2024-07-02 15:01:01,617 ERROR Job 7f85f10b-d3cd-4c47-a014-dc7746aed7e2 failed: 'ru'
['  File "D:\\ComfyUI_windows_portable\\ComfyUI\\execution.py", line 151, in recursive_execute\n    output_data, output_ui = get_output_data(obj, input_data_all)\n                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n', '  File "D:\\ComfyUI_windows_portable\\ComfyUI\\execution.py", line 81, in get_output_data\n    return_values = map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True)\n                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n', '  File "D:\\ComfyUI_windows_portable\\ComfyUI\\execution.py", line 74, in map_node_over_list\n    results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))\n                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n', '  File "D:\\ComfyUI_windows_portable\\ComfyUI\\custom_nodes\\ComfyUI_Custom_Nodes_AlekPet\\ArgosTranslateNode\\argos_translate_node.py", line 182, in argos_translate_text\n    self.langs_support = ALL_CODES[from_translate]["targets"]\n                         ~~~~~~~~~^^^^^^^^^^^^^^^^\n']
2024-07-02 15:01:06,234 ERROR Job 4d98eac8-ff8e-4212-b49d-152aa3560f4e failed: 'ru'
['  File "D:\\ComfyUI_windows_portable\\ComfyUI\\execution.py", line 151, in recursive_execute\n    output_data, output_ui = get_output_data(obj, input_data_all)\n                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n', '  File "D:\\ComfyUI_windows_portable\\ComfyUI\\execution.py", line 81, in get_output_data\n    return_values = map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True)\n                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n', '  File "D:\\ComfyUI_windows_portable\\ComfyUI\\execution.py", line 74, in map_node_over_list\n    results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))\n                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n', '  File "D:\\ComfyUI_windows_portable\\ComfyUI\\custom_nodes\\ComfyUI_Custom_Nodes_AlekPet\\ArgosTranslateNode\\argos_translate_node.py", line 182, in argos_translate_text\n    self.langs_support = ALL_CODES[from_translate]["targets"]\n                         ~~~~~~~~~^^^^^^^^^^^^^^^^\n']
2024-07-02 15:01:07,134 ERROR Job 5434b0a4-4711-41d7-9d17-bb39d234cae4 failed: 'ru'
['  File "D:\\ComfyUI_windows_portable\\ComfyUI\\execution.py", line 151, in recursive_execute\n    output_data, output_ui = get_output_data(obj, input_data_all)\n                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n', '  File "D:\\ComfyUI_windows_portable\\ComfyUI\\execution.py", line 81, in get_output_data\n    return_values = map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True)\n                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n', '  File "D:\\ComfyUI_windows_portable\\ComfyUI\\execution.py", line 74, in map_node_over_list\n    results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))\n                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n', '  File "D:\\ComfyUI_windows_portable\\ComfyUI\\custom_nodes\\ComfyUI_Custom_Nodes_AlekPet\\ArgosTranslateNode\\argos_translate_node.py", line 182, in argos_translate_text\n    self.langs_support = ALL_CODES[from_translate]["targets"]\n                         ~~~~~~~~~^^^^^^^^^^^^^^^^\n']
2024-07-02 15:01:07,601 ERROR Job 7315a369-be80-4521-91b9-8fbdc6caa148 failed: 'ru'
['  File "D:\\ComfyUI_windows_portable\\ComfyUI\\execution.py", line 151, in recursive_execute\n    output_data, output_ui = get_output_data(obj, input_data_all)\n                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n', '  File "D:\\ComfyUI_windows_portable\\ComfyUI\\execution.py", line 81, in get_output_data\n    return_values = map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True)\n                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n', '  File "D:\\ComfyUI_windows_portable\\ComfyUI\\execution.py", line 74, in map_node_over_list\n    results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))\n                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n', '  File "D:\\ComfyUI_windows_portable\\ComfyUI\\custom_nodes\\ComfyUI_Custom_Nodes_AlekPet\\ArgosTranslateNode\\argos_translate_node.py", line 182, in argos_translate_text\n    self.langs_support = ALL_CODES[from_translate]["targets"]\n                         ~~~~~~~~~^^^^^^^^^^^^^^^^\n']

`

**Reply by Drakosha405** [reply]
> Unfortunately mine doesn't work, could you please take a look ? 

I'm not familiar enough with the code to determine what caused the error. as you can see, everything worked for many people. 
I was just experimenting with it, nothing more.
try DeepTranslatorCLIPTextEncodeNode maybe its work for u.
`return self.add("DeepTranslatorCLIPTextEncodeNode", 1, clip=clip, text=text, from_translate="russian", auth_data=None)`

**Reply by XylitolJ** [reply]
I tried with DeepTranslator and it work! thank you

**Comment by dkqjhoifh** [comment]
Thank you very much, it worked!! I can finally type in Chinese!!

**Comment by lovemakerz** [comment]
Hello, awesome,
You can explain for beginners ?

Only edit this line ?
or another file / script is needed ?
( i'm french )

Thanks

**Reply by dkqjhoifh** [reply]
I made a video of the process and posted it for you to take a look at, and thank you very much Drakosha405 for your guidance.
https://www.bilibili.com/video/BV1Mp3ceTEPP/?spm_id_from=333.337.search-card.all.click&vd_source=8aaacb87515469df4606a33d29f31ec6

**Reply by Drakosha405** [reply]
> Hello, awesome, You can explain for beginners ?
> 
> Only edit this line ? or another file / script is needed ? ( i'm french )
> 
> Thanks

only this line edit. and u need [this](https://github.com/AlekPet/ComfyUI_Custom_Nodes_AlekPet) install
PS edit line will need to be done every time after updating this addon. save this information somewhere so that it is more convenient to replace it

**Reply by syl1130** [reply]
> > Hello, awesome, You can explain for beginners ?
> > Only edit this line ? or another file / script is needed ? ( i'm french )
> > Thanks
> 
> only this line edit. and u need [this](https://github.com/AlekPet/ComfyUI_Custom_Nodes_AlekPet) install PS edit line will need to be done every time after updating this addon. save this information somewhere so that it is more convenient to replace it

Hello Drakosha405, I referred to your method to modify the Deep translation, but there may be errors. I hope you can help me see how to modify the code when you have time. Thank you

`return self.add("DeepTranslatorCLIPTextEncodeNode", 1, clip=clip, text=text, from_translate="chinese simplified", service="MyMemoryTranslator [free]") `

**Reply by lovemakerz** [reply]
Thanks a lot, i have try, and working fine !!!

**Reply by Drakosha405** [reply]
> Hello Drakosha405, I referred to your method to modify the Deep translation, but there may be errors. I hope you can help me see how to modify the code when you have time. Thank you
> 
> `return self.add("DeepTranslatorCLIPTextEncodeNode", 1, clip=clip, text=text, from_translate="chinese simplified", service="MyMemoryTranslator [free]") `

I started it, and it seems to translate from the desired language, but when loading the workflow, the value from_translate = auto is always displayed for verification.
`return self.add("DeepTranslatorCLIPTextEncodeNode", 1, clip=clip, text=text, from_translate="russian", auth_data=None, service="MyMemoryTranslator [free]")`

but the check showed that in auto mode there is no translation of the mixed prompt, and when you select the desired one, everything is translated.


Prompt: cyberpunk, кирпичный дом  (cyberpunk, bricks house) with same seed.
Prompt: girl, кошачьи уши  (girl, cat ears) with same seed.

from_translate="auto"
![image](https://github.com/Acly/krita-ai-diffusion/assets/146111805/05576e4c-f758-4aac-a451-98f1e6e46d7c)
![image](https://github.com/Acly/krita-ai-diffusion/assets/146111805/a0dd45b8-c143-4626-bdbd-f0059ca1028a)


from_translate="russian"
![image](https://github.com/Acly/krita-ai-diffusion/assets/146111805/34fc7316-9994-47cf-8354-dc9f886ea480)
![image](https://github.com/Acly/krita-ai-diffusion/assets/146111805/93515233-3435-4685-a635-d664281c0641)

**Reply by Marksusu** [reply]
有这个翻译节点实在太好用了

**Reply by SyamsQ** [reply]
> > Hello Drakosha405, I referred to your method to modify the Deep translation, but there may be errors. I hope you can help me see how to modify the code when you have time. Thank you
> > `return self.add("DeepTranslatorCLIPTextEncodeNode", 1, clip=clip, text=text, from_translate="chinese simplified", service="MyMemoryTranslator [free]") `
> 
> I started it, and it seems to translate from the desired language, but when loading the workflow, the value from_translate = auto is always displayed for verification. `return self.add("DeepTranslatorCLIPTextEncodeNode", 1, clip=clip, text=text, from_translate="russian", auth_data=None, service="MyMemoryTranslator [free]")`
> 
> but the check showed that in auto mode there is no translation of the mixed prompt, and when you select the desired one, everything is translated.
> 
> Prompt: cyberpunk, кирпичный дом (cyberpunk, bricks house) with same seed. Prompt: girl, кошачьи уши (girl, cat ears) with same seed.
> 
> from_translate="auto" ![image](https://private-user-images.githubusercontent.com/146111805/343765259-05576e4c-f758-4aac-a451-98f1e6e46d7c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk2MDEyNTksIm5iZiI6MTcxOTYwMDk1OSwicGF0aCI6Ii8xNDYxMTE4MDUvMzQzNzY1MjU5LTA1NTc2ZTRjLWY3NTgtNGFhYy1hNDUxLTk4ZjFlNmU0NmQ3Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyOFQxODU1NTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lZGQyYjdkZWU2OGY2ZDQzODlkMGUzZTk4ZjkwMjNkZTFmZTI2YWY3YjQwYWJkNzJhNjBhZWQ2NDg4MWMwOTFmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.syxOmb-E5yaOAU8TlEyRfUdG1_rwLIqSDb0is-LoaIg) ![image](https://private-user-images.githubusercontent.com/146111805/343768604-a0dd45b8-c143-4626-bdbd-f0059ca1028a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk2MDEyNTksIm5iZiI6MTcxOTYwMDk1OSwicGF0aCI6Ii8xNDYxMTE4MDUvMzQzNzY4NjA0LWEwZGQ0NWI4LWMxNDMtNDYyNi1iZGJkLWYwMDU5Y2ExMDI4YS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyOFQxODU1NTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kNTQ5YmY0ODE1Mjg1NzRjNzJmMzc2M2MzMDg0OWNkNmVlZDk4MTYwZjM2ZmE3NzM5NDA0NjU1YzE4OTBmNWY4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.kwZIkX-vWOA32AZfUL_fnNCYkp9zJg26O_h3PfVnfQc)
> 
> from_translate="russian" ![image](https://private-user-images.githubusercontent.com/146111805/343765452-34fc7316-9994-47cf-8354-dc9f886ea480.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk2MDEyNTksIm5iZiI6MTcxOTYwMDk1OSwicGF0aCI6Ii8xNDYxMTE4MDUvMzQzNzY1NDUyLTM0ZmM3MzE2LTk5OTQtNDdjZi04MzU0LWRjOWY4ODZlYTQ4MC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyOFQxODU1NTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03ODkyMDRhZTliOTlhYTZiYTRmNjViNjU0ZTcxNzhmYmU4OGEwYmMzN2U0MzViNjNlZGU5NDk4Mzk4MWMzNDE2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.z_bUCFMFbqwi89KEYVz8931x47pCumVNmnLKvtbX9Lo) ![image](https://private-user-images.githubusercontent.com/146111805/343768688-93515233-3435-4685-a635-d664281c0641.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk2MDEyNTksIm5iZiI6MTcxOTYwMDk1OSwicGF0aCI6Ii8xNDYxMTE4MDUvMzQzNzY4Njg4LTkzNTE1MjMzLTM0MzUtNDY4NS1hNjM1LWQ2NjQyODFjMDY0MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyOFQxODU1NTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lZWM2OGY0ZGE5MTM2NDE1ZmU0NTViZWYwMmU4ZmI3MmY0M2U5YmE2ODUzYjI4Y2Q0ZDIwMWVjMWZkYTEyNjNjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.fjZ9znGspb2farVNCWzCmwUDWcAngMuALo3VQMiZqK0)

do we need something to download for get this translate feature?

**Reply by Drakosha405** [reply]
> do we need something to download for get this translate feature?

u need [this](https://github.com/AlekPet/ComfyUI_Custom_Nodes_AlekPet) install

**Comment by Drakosha405** [comment]
be careful when using DeepTranslatorCLIPTextEncodeNode encountered this. And it broke everything
`Error: Server Error: You made too many requests to the server.According to google, you are allowed to make 5 requests per secondand up to 200k requests per day. You can wait and try again later oryou can try the translate_batch function` 

I had to remove service="MyMemoryTranslator [free]" ... let's see if there will be such a mistake again.

**Reply by syl1130** [reply]
only remove ` , service="MyMemoryTranslator [free]"` ?

**Reply by Drakosha405** [reply]
> only remove ` , service="MyMemoryTranslator [free]"` ?

yes

**Comment by xingewh** [comment]
Please note that when using the Argos translation node, if you input English, it is highly likely that they will output incorrect results. Moreover, the actual translation results are not visible on the Krita AI interface.
![屏幕截图 2024-07-02 210730](https://github.com/Acly/krita-ai-diffusion/assets/57088247/edf65e81-492c-4616-ad7b-b24bd3059db8)

**Reply by Marksusu** [reply]
的确，应该有一个单独的“提示词翻译、生成、修改面板”（可以结合某些借助 ollama 的提示词生成工具）── 仅仅用于调整英文提示词的草稿，直到修改满意了，再应用到正反提示词节点。

**Reply by xingewh** [reply]
> 的确，应该有一个单独的“提示词翻译、生成、修改面板”（可以结合某些借助 ollama 的提示词生成工具）── 仅仅用于调整英文提示词的草稿，直到修改满意了，再应用到正反提示词节点。

是的，现在的Argos节点在翻译长文本的时候，结果有时不准确。我在翻译节点的那个项目里和作者就这一问题进行了沟通。这个不准备可能是因为ArgosTranslateCLIPTextEncodeNode使用的是本地模型进行翻译，模型规模较小，可能存在问题。DeepTranslatorCLIPTextEncodeNode就好很多，它是在线调用谷歌翻译等网站的API来实现的，但在国内需要使用全局代理才可以，并且就算是使用了全局代理，我把它加到了kirta的代码中也会出现与网络相关的报错。

为了准确的使用结果，还是使用原来的英文输入吧。可以在comfyui的上建一个合适的翻译工作流，然后复制过来。

**Reply by Marksusu** [reply]
> > 的确，应该有一个单独的“提示词翻译、生成、修改面板”（可以结合某些借助 ollama 的提示词生成工具）── 仅仅用于调整英文提示词的草稿，直到修改满意了，再应用到正反提示词节点。
> 
> 是的，现在的Argos节点在翻译长文本的时候，结果有时不准确。我在翻译节点的那个项目里和作者就这一问题进行了沟通。这个不准备可能是因为ArgosTranslateCLIPTextEncodeNode使用的是本地模型进行翻译，模型规模较小，可能存在问题。DeepTranslatorCLIPTextEncodeNode就好很多，它是在线调用谷歌翻译等网站的API来实现的，但在国内需要使用全局代理才可以，并且就算是使用了全局代理，我把它加到了kirta的代码中也会出现与网络相关的报错。
> 
> 为了准确地使用结果，还是使用原来的英文输入吧。可以在comfyui的上建一个合适的翻译工作流，然后复制过来。

`https://github.com/kingzcheung/ComfyUI_kkTranslator_nodes`
用这个改一下是可以的，kkTranslator 是用的本地模型，能忠实地翻译逐个短语，比 Ollama 稳定（Ollama 里有些模型会胡言乱语，但好处是你可以指派它去编写提示词，任其发挥，也很有意思：`https://github.com/ScreamingHawk/comfyui-ollama-prompt-encode`，也值得一试。）。
我还发现，如果先在终端打开 ComfyUI， 再连接 Krita 就能在终端看到每次的翻译结果，这真是太好玩了。

**Comment by Drakosha405** [comment]
in 1.20.0 version.
now we need edit 355 line in comfy_workflow.py

**Comment by Acly** [comment]
Localisation stuff for next version:

![Screenshot 2024-07-23 180143](https://github.com/user-attachments/assets/52941401-c03e-48be-9192-b3c1e17d543c)

I'm using the argos-translate models. They're okay. Google is better, but I don't really want to deal with bad or blocked connections, rate limiting, api keys...

![Screenshot 2024-07-23 181448](https://github.com/user-attachments/assets/1fd23de0-26ad-4352-950e-7b1e553fa3f5)

The language indicator is clickable to toggle on/off, or with Ctrl+Click it will translate & replace the prompt directly.

**Reply by Marksusu** [reply]
我让 ChatGPT 帮我写代码，用 kkTranslator 改了一个，这个翻译模型效果也不错😀

[https://github.com/Marksusu/ComfyUI_MTCLIPEncode](https://github.com/Marksusu/ComfyUI_MTCLIPEncode)

**Reply by Acly** [reply]
It would be better to never translate the style prompt?

**Reply by Marksusu** [reply]
我是这样理解的：依赖母语输入只是避免思路中断的临时举措。

当你修改、添加某个母语词汇时，翻译工具会把所有提示词视为一个整体，全部重新翻译一次，它会“智能地”判断语境、调整用词，致使翻译结果不稳定。

如果我们希望结果可控、希望“上次起作用的提示词”保持不变，就要及时改为英语，并指明翻译工具不要动这一部分。

翻译工具还应该把对照的翻译结果展示给人类，这样我们才能了解 ── 具体是哪些英文词汇真正进入了编码流程，以及 `_:.()[]<>` 之类的符号有没有被遗漏或被改变。

**Reply by Marksusu** [reply]
传统翻译工具有时会显得愚笨、太机械化、不能理解人类的意图；
而智能模型会对人类意图做不必要的揣测，会对一连串的负面提示词感到惶恐，陷入道德危机，进而拒绝翻译。
它会说：
“我不知道你在说什么……”
“我不能那样做……”
“我们无法完成这个任务……”

**Reply by syl1130** [reply]
@Acly Looking forward to the update, I can't wait to use it

**Comment by Drakosha405** [comment]
@Acly 
in wiki to translate.
`To update an existing translation (eg. after the plugin has been updated and new text was added) simply search for entries which are null. These are valid, but not not translated yet.`

maybe it's worth adding all new words to the end of the list, despite the alphabetical order.  because it will be very difficult for everyone to translate new words looking for them every time in the middle. something like this

"translations": {
|--"packages available": null,
|--"packages require installation": null,
|--"packages selected": null
v.1.22
|--"new word" : null
|--"new word" : null
|--"new word" : null
v.1.23
|--"new word" : null
|--"new word" : null
|--"new word" : null
}

**Reply by Acly** [reply]
Maybe. But Ctrl+F search for null is not that difficult? Even notepad can do that.

The translation list is auto-generated, and it also needs to track when existing text changes or is removed. I haven't actually implemented the update yet, but there will be a way.

**Reply by Drakosha405** [reply]
> Maybe. But Ctrl+F search for null is not that difficult? Even notepad can do that.
> 
> The translation list is auto-generated, and it also needs to track when existing text changes or is removed. I haven't actually implemented the update yet, but there will be a way.

the question is whether all language files will be updated automatically or only template. if it's all at once, then it's still okay.  if only template, it will become difficult to add new positions to files of other localizations.

**Reply by Marksusu** [reply]
@Acly 大佬，要不……干脆…… 在 Krita 里面加一个 Ollama 的聊天窗口吧。

这样一来，人类就可以跟自己的本地模型商量如何翻译、推敲用哪个英文字眼更准确。

人类用母语或者任何什么语言描述自己脑中的想象图画，Ollama 为人类提供英文 prompt 建议，再由人类自由决定最终填写到提示词输入框里的内容到底是什么。

这样，“构思提示词”和“生成图像”就成了十分清晰的两个步骤。

另外，这样做还有个好处，Ollama 可以借助 Llava 这类视觉模型“看到”我们画布里的内容，把图中的事物描述为英文文本，相当于在做 prompt 与图像的相互验证。

这就等于有了多个勤快的翻译、顾问、同事、书童……与我们一起头脑风暴，这恐怕才是“人类携手 AI 共同迈向未来”的优雅姿势😅😂😂

**Reply by Marksusu** [reply]
> @Acly 大佬，要不……干脆…… 在 Krita 里面加一个 Ollama 的聊天窗口吧。
> 
> blabla

https://github.com/Acly/krita-ai-diffusion/discussions/1133

https://github.com/Marksusu/ComfyUI_MTCLIPEncode

Ollama 有了，但只能优化提示词。
这个节点是 GPT 帮我写的，“实现聊天机器人”什么的，对我来说太难了，完全不懂。

**Reply by syl1130** [reply]
@Marksusu  建议你用英文语言发给Acly吧 中文他看不懂的

**Reply by Marksusu** [reply]
> @Marksusu  建议你用英文语言发给Acly吧 中文他看不懂的

没关系，他会用翻译😀

**Reply by syl1130** [reply]
你是认真的吗

**Reply by Marksusu** [reply]
> 你是认真的吗

哈哈，是的，谢谢你关心我～

我是认真的，

他要是有时间、有心情时，的确会用翻译工具读别人写的东西。

其实，我也曾用翻译工具翻译成英文发过几个帖子，但实际效果并不好，实在没必要那样做。

毕竟用自己的母语才能保证准确地表达我之所想。若是谁用翻译工具译出歧义，那是工具的责任，但翻译工具会进步，并且进步会比我快，早晚能准确还原我发的信息，而我最初用母语做出的表达是不会变的。

在我有自信在某特定场景直接使用某种外语前，我还是更倾向使用我能信任的母语来表达我的想法，这才是目前我能负担得起的责任。也是目前我自己能认可的对交流双方最低消耗的办法。

至少在半年前，我还不是这种态度，但当我发现大模型会把“草”翻译成 “F” 打头的那个词的时候，我决定非必要情况只用母语来发帖子。
